doc/design-terrafor-provider.rst

```diff
@@ -0,0 +1,175 @@
+==================================
+Ganeti Terraform Provider Proposal
+==================================
+
+:Created: 2023-Feb-23
+:Status: Draft
+
+.. toctree::
+   :hidden:
+
+1. Introduction
+---------------
+
+This document proposes the creation of a Terraform provider for Ganeti, 
+a cluster virtualization management software. The provider will 
+allow users to manage their Ganeti infrastructure as code, 
+using the Terraform configuration language and tool. 
+The provider will be implemented using the Terraform provider SDK 
+and the Ganeti Remote API.
+
+2. Background
+-------------
+
+Terraform is an infrastructure-as-code tool that enables users to manage
+and provision cloud infrastructure using a declarative configuration 
+language. Terraform is platform-agnostic and supports a wide range of 
+infrastructure providers, including major cloud providers such as AWS, 
+Azure, and Google Cloud. 
+Terraform providers are plugins that extend Terraform's capabilities 
+to support new infrastructure platforms or services.
+
+Ganeti is a cluster virtualization management software that allows users
+to create and manage virtual machines (VMs) on a cluster of 
+physical hosts. Ganeti provides a number of features, such as automatic 
+failover, live migration, and DRBD-based disk replication. 
+Ganeti also has a Remote API that enables users to manage their Ganeti 
+cluster programmatically.
+
+3. Scope
```

Comment:
This section should go a bit more into detail about what should be implemented or can be expected from the provider, so that we end up with something people can use to create code and/or also use to compare against PRs etc.

- supported hypervisors
- supported disk templates
- supported disk schemes (number of disks etc.)
- supported network schemes (numbers of nics, network types etc.)
- other supported features (e.g. automatic/manual node selection, custom hypervisor parameters)

I would suggest to start simple:
- we only support KVM for the first iteration
- all possible disk templates
- only automatic instance allocation
- any number of disks, any number of NICs
- network mode `bridged` with optional parameters `link`, `network` and `vlan` (this should cover most usecases)
- all backend parameters (`maxmem`, `minmem`, `vcpus`)
- all current KVM hvparams (given that we are using a strongly typed language, supporting "any" parameter will be harder to implement. OTOH most people will likely use only few hvparams on instance level as most will usually be set on cluster level)
- set OS type (and possibly also OS parameters?)

We also need to agree on what the provider should do with instance changes:
- will new disks/NICs be hot- or cold-added? Will this be configurable?
- will disks/NICs be deleted? Do we support that at all? Or maybe only in later releases? How should the terraform data be modeled to "detect" that a disk/NIC has been removed from the instance definition?
- it should support disk growing
- what happens when backend/hv parameters are changed? will this trigger an automatic reboot of the instance? should this be configurable?

Reply:
@rbott i will incorporate your suggestion, however i still think i should be less as a start to not over-complicate it.

Reply:
This document needs to be fairly exhaustive to not over-complicate the development of the terraform provider itself :-) If we expect people to contribute, everyone needs to be able to understand planned features/goals/ideas. Otherwise we scare contributors away be rejecting their contributions because they do not meet whatever we might have in mind.

This document can/should also be used to decide whether the provider has reached a stable version 1.0 or if it is still missing features considered a bare minimum requirement.